pkg/kv/kvserver/replica_raftstorage_log.go

```diff
@@ -0,0 +1,325 @@
+		return nil, errors.Errorf("there is a gap in the index record between lo:%d and hi:%d at index:%d", lo, hi, expectedIndex)
+	}
+
+	// No results, was it due to unavailability or truncation?
+	ts, err := rsl.LoadRaftTruncatedState(ctx, reader)
+	if err != nil {
+		return nil, err
+	}
+	if ts.Index >= lo {
+		// The requested lo index has already been truncated.
+		return nil, raft.ErrCompacted
+	}
+	// The requested lo index does not yet exist.
+	return nil, raft.ErrUnavailable
+}
+
+// iterateEntries iterates over each of the Raft log entries in the range
+// [lo,hi). At each step of the iteration, f() is invoked with the current log
+// entry.
+//
+// The function does not accept a maximum number of entries or bytes. Instead,
+// callers should enforce any limits by returning iterutil.StopIteration from
+// the iteration function to terminate iteration early, if necessary.
+func iterateEntries(
+	ctx context.Context,
+	reader storage.Reader,
+	rangeID roachpb.RangeID,
+	lo, hi uint64,
+	f func(raftpb.Entry) error,
+) error {
+	key := keys.RaftLogKey(rangeID, lo)
+	endKey := keys.RaftLogKey(rangeID, hi)
+	iter := reader.NewMVCCIterator(storage.MVCCKeyIterKind, storage.IterOptions{
+		UpperBound: endKey,
+	})
+	defer iter.Close()
+
+	var meta enginepb.MVCCMetadata
+	var ent raftpb.Entry
+
+	iter.SeekGE(storage.MakeMVCCMetadataKey(key))
+	for ; ; iter.Next() {
+		if ok, err := iter.Valid(); err != nil || !ok {
+			return err
+		}
+
+		if err := protoutil.Unmarshal(iter.UnsafeValue(), &meta); err != nil {
+			return errors.Wrap(err, "unable to decode MVCCMetadata")
+		}
+		if err := storage.MakeValue(meta).GetProto(&ent); err != nil {
+			return errors.Wrap(err, "unable to unmarshal raft Entry")
+		}
+		if err := f(ent); err != nil {
+			return iterutil.Map(err)
+		}
+	}
+}
+
+// invalidLastTerm is an out-of-band value for r.mu.lastTerm that
+// invalidates lastTerm caching and forces retrieval of Term(lastTerm)
+// from the raftEntryCache/RocksDB.
+const invalidLastTerm = 0
+
+// Term implements the raft.Storage interface.
+// Term requires that r.mu is held for writing because it requires exclusive
+// access to r.mu.stateLoader.
+func (r *replicaRaftStorage) Term(i uint64) (uint64, error) {
+	// TODO(nvanbenschoten): should we set r.mu.lastTerm when
+	//   r.mu.lastIndex == i && r.mu.lastTerm == invalidLastTerm?
+	if r.mu.lastIndex == i && r.mu.lastTerm != invalidLastTerm {
+		return r.mu.lastTerm, nil
+	}
+	// Try to retrieve the term for the desired entry from the entry cache.
+	if e, ok := r.store.raftEntryCache.Get(r.RangeID, i); ok {
+		return e.Term, nil
+	}
+	readonly := r.store.Engine().NewReadOnly(storage.StandardDurability)
+	defer readonly.Close()
+	ctx := r.AnnotateCtx(context.TODO())
+	return term(ctx, r.mu.stateLoader, readonly, r.RangeID, r.store.raftEntryCache, i)
+}
+
+// raftTermLocked requires that r.mu is locked for writing.
+func (r *Replica) raftTermLocked(i uint64) (uint64, error) {
```

Comment:
I would leave all the `*Replica`-anchored methods where they are, so that we keep the `raft.Storage` impl in the original file. What we're moving out are `entries`, `term`, `iterateEntries`.

I would then separately tackle getting a `raft.Storage` implementation without a `*Replica` dependency. I think this might need some unpeeling and we need to make sure it's the fight we want to pick now (i.e. does it help with sep raft log). Let's talk about it separately.